test/extensions/filters/http/original_src/original_src_test.cc

```diff
@@ -0,0 +1,217 @@
+
+#include "test/mocks/buffer/mocks.h"
+#include "test/mocks/common.h"
+#include "test/mocks/http/mocks.h"
+#include "test/mocks/network/mocks.h"
+#include "test/test_common/printers.h"
+#include "test/test_common/utility.h"
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+using testing::_;
+using testing::Exactly;
+using testing::SaveArg;
+using testing::StrictMock;
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace OriginalSrc {
+namespace {
+
+class OriginalSrcHttpTest : public testing::Test {
+public:
+  std::unique_ptr<OriginalSrcFilter> makeDefaultFilter() {
+    return makeFilterWithCallbacks(callbacks_);
+  }
+
+  std::unique_ptr<OriginalSrcFilter>
+  makeFilterWithCallbacks(Http::StreamDecoderFilterCallbacks& callbacks) {
+    Config default_config;
+    auto filter = std::make_unique<OriginalSrcFilter>(default_config);
+    filter->setDecoderFilterCallbacks(callbacks);
+    return filter;
+  }
+
+  std::unique_ptr<OriginalSrcFilter> makeMarkingFilter(uint32_t mark) {
+    envoy::config::filter::http::original_src::v2alpha1::OriginalSrc proto_config;
+    proto_config.set_mark(mark);
+
+    Config config(proto_config);
+    auto filter = std::make_unique<OriginalSrcFilter>(config);
+    filter->setDecoderFilterCallbacks(callbacks_);
+    return filter;
+  }
+
+  void setAddressToReturn(const std::string& address) {
+    callbacks_.stream_info_.downstream_remote_address_ = Network::Utility::resolveUrl(address);
+  }
+
+protected:
+  StrictMock<MockBuffer> buffer_;
+  NiceMock<Http::MockStreamDecoderFilterCallbacks> callbacks_;
+  NiceMock<Network::MockConnectionSocket> socket_;
+  Http::TestHeaderMapImpl headers_;
+
+  absl::optional<Network::Socket::Option::Details>
+  findOptionDetails(const Network::Socket::Options& options, Network::SocketOptionName name,
+                    envoy::api::v2::core::SocketOption::SocketState state) {
+    for (const auto& option : options) {
+      auto details = option->getOptionDetails(socket_, state);
+      if (details.has_value() && details->name_ == name) {
+        return details;
+      }
+    }
+
+    return absl::nullopt;
+  }
+};
+
+TEST_F(OriginalSrcHttpTest, OnNonIpAddressDecodeSkips) {
+  auto filter = makeDefaultFilter();
+  setAddressToReturn("unix://domain.socket");
+  EXPECT_CALL(callbacks_, addUpstreamSocketOptions(_)).Times(0);
+  EXPECT_EQ(filter->decodeHeaders(headers_, false), Http::FilterHeadersStatus::Continue);
+}
+
+TEST_F(OriginalSrcHttpTest, DecodeHeadersIpv4AddressAddsOption) {
+  auto filter = makeDefaultFilter();
+
+  Network::Socket::OptionsSharedPtr options;
+  setAddressToReturn("tcp://1.2.3.4:0");
+  EXPECT_CALL(callbacks_, addUpstreamSocketOptions(_)).WillOnce(SaveArg<0>(&options));
+
+  EXPECT_EQ(filter->decodeHeaders(headers_, false), Http::FilterHeadersStatus::Continue);
+
+  // Not ideal -- we're assuming that the original_src option is first, but it's a fair assumption
```

Comment:
Is there something preventing you from making a better assertion here? ie would iterating over the options and verifying that the original_src option is present be possible?

Reply:
I guess I could use `dynamic_cast` to search for it. Unfortunately, the introspection stuff we added doesn't work well  for options which don't actually map to a linux socket option (the lookup is by 'name', which is a pair of ints).

Reply:
I've decided to just change it to assert on the behaviour. I don't like control flow in UT (again, to keep it simple), so I was trying to avoid looping, but I think I overdid it here. The goal was to keep it simple, but I was making assumptions that actually made the test more fragile, so a loop applying the option is a better solution.